frame/contracts/src/wasm/mod.rs

```diff
@@ -41,81 +37,61 @@ pub use crate::wasm::{
 
 use crate::{
 	exec::{ExecResult, Executable, ExportedFunction, Ext},
-	gas::GasMeter,
-	AccountIdOf, BalanceOf, CodeHash, CodeVec, Config, Error, OwnerInfoOf, RelaxedCodeVec,
-	Schedule, LOG_TARGET,
+	gas::{GasMeter, Token},
+	weights::WeightInfo,
+	AccountIdOf, BadOrigin, BalanceOf, CodeHash, CodeInfoOf, CodeVec, Config, Error, Event, Pallet,
+	PristineCode, Schedule, Weight, LOG_TARGET,
 };
 use codec::{Decode, Encode, MaxEncodedLen};
-use frame_support::dispatch::{DispatchError, DispatchResult};
+use frame_support::{
+	dispatch::{DispatchError, DispatchResult},
+	ensure,
+	traits::ReservableCurrency,
+};
 use sp_core::Get;
-use sp_runtime::RuntimeDebug;
+use sp_runtime::{traits::Hash, RuntimeDebug};
 use sp_std::prelude::*;
 use wasmi::{
-	Config as WasmiConfig, Engine, Instance, Linker, Memory, MemoryType, Module, StackLimits, Store,
+	Config as WasmiConfig, Engine, FuelConsumptionMode, Instance, Linker, Memory, MemoryType,
+	Module, StackLimits, Store,
 };
 
-/// A prepared wasm module ready for execution.
-///
-/// # Note
-///
-/// This data structure is mostly immutable once created and stored. The exceptions that
-/// can be changed by calling a contract are `instruction_weights_version` and `code`.
-/// `instruction_weights_version` and `code` change when a contract with an outdated instrumentation
-/// is called. Therefore one must be careful when holding any in-memory representation of this
-/// type while calling into a contract as those fields can get out of date.
-#[derive(Clone, Encode, Decode, scale_info::TypeInfo, MaxEncodedLen)]
+/// Validated Wasm module ready for execution.
+/// This data structure is immutable once created and stored.
+#[derive(Encode, Decode, scale_info::TypeInfo)]
 #[codec(mel_bound())]
 #[scale_info(skip_type_params(T))]
-pub struct PrefabWasmModule<T: Config> {
-	/// Version of the instruction weights with which the code was instrumented.
-	#[codec(compact)]
-	instruction_weights_version: u32,
-	/// Initial memory size of a contract's sandbox.
-	#[codec(compact)]
-	initial: u32,
-	/// The maximum memory size of a contract's sandbox.
-	#[codec(compact)]
-	maximum: u32,
-	/// Code instrumented with the latest schedule.
-	code: RelaxedCodeVec<T>,
-	/// A code that might contain non deterministic features and is therefore never allowed
-	/// to be run on chain. Specifically this code can never be instantiated into a contract
-	/// and can just be used through a delegate call.
-	determinism: Determinism,
-	/// The uninstrumented, pristine version of the code.
-	///
-	/// It is not stored because the pristine code has its own storage item. The value
-	/// is only `Some` when this module was created from an `original_code` and `None` if
-	/// it was loaded from storage.
```

Comment:
why moving this to CodeInfo, don't we need it every time we load the code?

Reply:
it [was](https://github.com/paritytech/substrate/pull/14084#discussion_r1193095986) decided to store code-related "_metadata_" separately from code. But I get your point here, `determinism` field is just 1 byte, while loading the whole `CodeInfo` will read a couple of dozens of bytes, at the same time `deposit` and `owner` are not needed every time we load the code. Hence there is a room for a little optimization here. @athei wdyt?

Reply:
> while loading the whole CodeInfo will read a couple of dozens of bytes
and it's another io operation to the db you have to pay for

Reply:
yes